src/middleware/compression.rs

```diff
@@ -0,0 +1,434 @@
+    }
+
+    fn preferred_encoding(&self, headers: &HeaderMap) -> Result<Encoding, Error> {
+        let encoding = match accept_encoding::parse(headers) {
+            Ok(encoding) => encoding,
+            Err(e) => return Err(Error::from(StatusCode::BAD_REQUEST)),
+        };
+        match encoding {
+            Some(encoding) => Ok(encoding),
+            None => Ok(self.default_encoding),
+        }
+    }
+
+    /// Consumes the response and returns an encoded version of it.
+    fn encode(&self, mut res: Response, encoding: Encoding) -> Response {
+        // TODO: test and make sure this works
+        match encoding {
+            Encoding::Gzip => {
+                let mut body = Body::empty();
+                std::mem::swap(&mut body, res.body_mut());
+                let stream = stream::GzipEncoder::new(body, self.settings.gzip);
+                *res.body_mut() = Body::from_stream(stream);
+                res.headers_mut()
+                    .append(CONTENT_ENCODING, encoding.to_header_value());
+                res
+            }
+            Encoding::Deflate => {
+                let mut body = Body::empty();
+                std::mem::swap(&mut body, res.body_mut());
+                let stream = stream::ZlibEncoder::new(body, self.settings.deflate);
+                *res.body_mut() = Body::from_stream(stream);
+                res.headers_mut()
+                    .append(CONTENT_ENCODING, encoding.to_header_value());
+                res
+            }
+            Encoding::Brotli => {
+                let mut body = Body::empty();
+                std::mem::swap(&mut body, res.body_mut());
+                let stream = stream::BrotliEncoder::new(body, self.settings.brotli);
+                *res.body_mut() = Body::from_stream(stream);
+                res.headers_mut()
+                    .append(CONTENT_ENCODING, encoding.to_header_value());
+                res
+            }
+            Encoding::Zstd => {
+                let mut body = Body::empty();
+                std::mem::swap(&mut body, res.body_mut());
+                let stream = stream::ZstdEncoder::new(body, self.settings.zstd);
+                *res.body_mut() = Body::from_stream(stream);
+                res.headers_mut()
+                    .append(CONTENT_ENCODING, encoding.to_header_value());
+                res
+            }
+            Encoding::Identity => res,
+        }
+    }
+}
+
+impl<Data: Send + Sync + 'static> Middleware<Data> for Compression {
+    fn handle<'a>(&'a self, cx: Context<Data>, next: Next<'a, Data>) -> BoxFuture<'a, Response> {
+        box_async! {
+            let encoding = match self.preferred_encoding(cx.headers()) {
+                Ok(encoding) => encoding,
+                Err(e) => return e.into_response(),
+            };
+            let res = next.run(cx).await;
+            self.encode(res, encoding)
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+enum Decoding {
```

Comment:
Can't we just use `Encoding` for this? (I'm not quite familiar with `Content-Encoding` and `Accept-Encoding` things)

Reply:
We can -- I had to include this separately because of the existence of `Encoding::Any` previously, but I should be able to simplify this now.